mlx/backend/cuda/allocator.cpp

```diff
@@ -88,15 +90,19 @@ CudaAllocator::CudaAllocator()
           page_size,
           [](CudaBuffer* buf) { return buf->size; },
           [this](CudaBuffer* buf) { cuda_free(buf); }) {
-  // TODO: Set memory limit for multi-device.
   size_t free, total;
   CHECK_CUDA_ERROR(cudaMemGetInfo(&free, &total));
   memory_limit_ = total * 0.95;
   max_pool_size_ = memory_limit_;
+  int loc = 0;
+  CHECK_CUDA_ERROR(cudaDeviceGetDefaultMemPool(&cuda_pool_, loc));
+  CHECK_CUDA_ERROR(cudaMemPoolSetAttribute(
+      cuda_pool_, cudaMemPoolAttrReleaseThreshold, &memory_limit_));
```

Comment:
The MLX buffer cache and CUDA memory pool cache does not work together: MLX wouldn't free buffers until `memory_limit_` is exceeded so CUDA memory pool does not get a chance to reuse memory. The only case that CUDA's memory pool cache kicks in is after `clear_cache()` gets called, which I think is against the purpose of `clear_cache()`.

I think we should only use one kind of buffer cache at the same time, probably depends on which one is faster.